File: src/go/shell/vfs/local_folder_source.go
package vfs

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// LocalFolderSource mirrors a host directory as a read-only MountSource.
//
// Path-escape-safe: the real absolute root is resolved once, and every subpath
// is joined, resolved (following symlinks), and asserted to live at or under
// the root — defeating "../" and symlink escapes. Absolute subpaths are
// rejected; out-of-root paths are treated as not-found.
type LocalFolderSource struct {
	root string
}

// NewLocalFolderSource constructs a source rooted at the given host directory.
func NewLocalFolderSource(root string) *LocalFolderSource {
	resolved, err := filepath.EvalSymlinks(root)
	if err != nil {
		abs, aerr := filepath.Abs(root)
		if aerr != nil {
			resolved = filepath.Clean(root)
		} else {
			resolved = abs
		}
	}
	return &LocalFolderSource{root: resolved}
}

// resolve maps a relative subpath onto a confined host path, or returns
// ("", false) if it would escape the root or is absolute.
func (s *LocalFolderSource) resolve(subpath string) (string, bool) {
	if filepath.IsAbs(subpath) {
		return "", false
	}
	joined := filepath.Join(s.root, filepath.FromSlash(subpath))
	resolved, err := filepath.EvalSymlinks(joined)
	if err != nil {
		// Path may not exist yet — fall back to the lexically cleaned join so
		// not-found is distinguishable from escape.
		resolved = filepath.Clean(joined)
	}
	if resolved != s.root && !strings.HasPrefix(resolved, s.root+string(os.PathSeparator)) {
		return "", false
	}
	return resolved, true
}

func (s *LocalFolderSource) Stat(subpath string) (MountStat, bool) {
	resolved, ok := s.resolve(subpath)
	if !ok {
		return MountStat{}, false
	}
	info, err := os.Stat(resolved)
	if err != nil {
		return MountStat{}, false
	}
	return MountStat{
		IsDir: info.IsDir(),
		Size:  info.Size(),
		Mtime: info.ModTime().Unix(),
	}, true
}

func (s *LocalFolderSource) List(subpath string) ([]string, error) {
	resolved, ok := s.resolve(subpath)
	if !ok {
		return nil, os.ErrNotExist
	}
	entries, err := os.ReadDir(resolved)
	if err != nil {
		return nil, err
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		names = append(names, e.Name())
	}
	sort.Strings(names)
	return names, nil
}

func (s *LocalFolderSource) Read(subpath string) ([]byte, error) {
	resolved, ok := s.resolve(subpath)
	if !ok {
		return nil, os.ErrNotExist
	}
	return os.ReadFile(resolved)
}

File: src/php/LocalFolderSource.php
<?php

declare(strict_types=1);

namespace AgentHarness;

/**
 * Mirrors a host directory as a read-only mount source.
 *
 * Path-escape-safe: the real absolute root is resolved once, and every subpath
 * is joined, resolved (following symlinks), and asserted to live at or under
 * the root -- defeating '../' and symlink escapes. Absolute subpaths are
 * rejected; out-of-root paths are treated as not-found.
 */
class LocalFolderSource implements MountSource
{
    private string $root;

    public function __construct(string $root)
    {
        $real = realpath($root);
        $this->root = $real !== false ? $real : $root;
    }

    private function resolve(string $subpath): ?string
    {
        // Reject absolute subpaths (POSIX '/...' or Windows 'C:\...').
        if ($subpath !== '' && ($subpath[0] === '/' || preg_match('#^[A-Za-z]:[\\\\/]#', $subpath) === 1)) {
            return null;
        }
        $joined = $this->root . DIRECTORY_SEPARATOR . $subpath;
        $real = realpath($joined);
        if ($real === false) {
            return null; // not found
        }
        if ($real !== $this->root && !str_starts_with($real, $this->root . DIRECTORY_SEPARATOR)) {
            return null; // escaped the root
        }
        return $real;
    }

    public function stat(string $subpath): ?MountStat
    {
        $resolved = $this->resolve($subpath);
        if ($resolved === null) {
            return null;
        }
        return new MountStat(
            isDir: is_dir($resolved),
            size: is_file($resolved) ? (int) filesize($resolved) : 0,
            mtime: (int) filemtime($resolved),
        );
    }

    /** @return list<string> */
    public function list(string $subpath): array
    {
        $resolved = $this->resolve($subpath);
        if ($resolved === null) {
            throw new \RuntimeException("{$subpath}: No such file");
        }
        $entries = scandir($resolved);
        if ($entries === false) {
            throw new \RuntimeException("{$subpath}: cannot list");
        }
        $names = array_values(array_filter($entries, fn(string $n) => $n !== '.' && $n !== '..'));
        sort($names);
        return $names;
    }

    public function read(string $subpath): string
    {
        $resolved = $this->resolve($subpath);
        if ($resolved === null) {
            throw new \RuntimeException("{$subpath}: No such file");
        }
        $content = file_get_contents($resolved);
        if ($content === false) {
            throw new \RuntimeException("{$subpath}: cannot read");
        }
        return $content;
    }
}
